services/mlmodel/tflitecpu/tflite_cpu.go

```diff
@@ -0,0 +1,249 @@
+		if params == nil {
+			return nil, errors.New("could not find parameters")
+		}
+		if params.NumThreads <= 0 {
+			loader, err = inf.NewDefaultTFLiteModelLoader()
+		} else {
+			loader, err = inf.NewTFLiteModelLoader(params.NumThreads)
+		}
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get loader")
+		}
+
+		fullpath, err2 := fp.Abs(params.ModelPath)
+		if err2 != nil {
+			model, err = loader.Load(params.ModelPath)
+		} else {
+			model, err = loader.Load(fullpath)
+		}
+
+		if err != nil {
+			if strings.Contains(err.Error(), "failed to load") {
+				if err2 != nil {
+					return nil, errors.Wrapf(err, "file not found at %s", params.ModelPath)
+				}
+				return nil, errors.Wrapf(err, "file not found at %s", fullpath)
+			}
+			return nil, errors.Wrap(err, "loader could not load model")
+		}
+		return model, nil
+	}
+	model, err = addModel()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not add model from location %s", params.ModelPath)
+	}
+
+	return &Model{attrs: *params, model: model, name: name}, nil
+}
+
+// Infer takes the input map, finds the image (labeled 'image'), and uses the inference
+// package to return the result from the tflite cpu model as a map.
+func (m *Model) Infer(ctx context.Context, input map[string]interface{}) (map[string]interface{}, error) {
+	_, span := trace.StartSpan(ctx, "service::mlmodel::tflite_cpu::Infer")
+	defer span.End()
+
+	outMap := make(map[string]interface{})
+	doInfer := func(input interface{}) (map[string]interface{}, error) {
+		outTensors, err := m.model.Infer(input)
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't infer from model %s", m.name)
+		}
+		// Fill in the output map with the names from metadata if u have them
+		// Otherwise, do output1, output2, etc.
+		for i := 0; i < len(m.model.Info.OutputTensorTypes); i++ {
+			if m.metadata != nil {
+				outMap[m.metadata.Outputs[i].Name] = outTensors[i]
+			} else {
+				outMap["output"+strconv.Itoa(i)] = outTensors[i]
+			}
+		}
+		return outMap, nil
+	}
+
+	// If there's only one thing in the input map, use it.
+	if len(input) == 1 {
+		for _, in := range input {
+			return doInfer(in)
+		}
+	}
+	// If you have more than 1 thing
+	if m.metadata != nil {
+		// Use metadata if possible to grab input name
+		return doInfer(input[m.metadata.Inputs[0].Name])
+	}
+	// Try to use "input"
+	if in, ok := input["input"]; ok {
+		return doInfer(in)
+	}
+
+	return nil, errors.New("input map has multiple elements and none are named 'input'")
+}
+
+// Metadata reads the metadata from your tflite cpu model into the metadata struct
+// that we use for the mlmodel service.
+func (m *Model) Metadata(ctx context.Context) (mlmodel.MLMetadata, error) {
+	_, span := trace.StartSpan(ctx, "service::mlmodel::tflite_cpu::Metadata")
+	defer span.End()
+
+	if m.metadata != nil {
+		return *m.metadata, nil
+	}
+
+	// model.Metadata() and funnel it into this struct
+	md, err := m.model.Metadata()
```

Comment:
Wanted to follow up and make sure this isn't relying on the inner metadata of the tflite model, as in the training pipeline we just convert a [TF model into TFLite](https://github.com/viamrobotics/app/blob/main/mltraining/python/ml_python_scripts/trainer/task.py#L152) and then supply the corresponding labels file, rather than embedding anyway metadata [this way](https://www.tensorflow.org/lite/models/convert/metadata).

Reply:
Hmmmm, this won't be a problem here, but will be a problem down the line for the vision service, because the metadata is where the Vision Service will figure out what the shape of the input and output tensors are. Without the meta data, the model does its "best guess" which is prone to fail.

Reply:
But either way, we will take into account the models made specifically by Viam in the new vision service, even with no model metadata

Reply:
Got it, thanks for explaining! Agreed that it sounds like we can cross that bridge when we come to it; created a backlog ticket to consider embedding metadata directly into our Viam-made tflite files.